### ISSUE TITLE
Backward engine computes unnecessary dependencies

### ISSUE THREAD

**Issue by apaszke** [pytorch/pytorch#726]
```python
import torch
from torch.autograd import Variable

x = Variable(torch.randn(5, 5), requires_grad=True)
y = Variable(torch.randn(5, 5), requires_grad=True)

a = x + y
b = torch.max(a, 1)[1].repeat(1, 5).float()
o = (b + a).sum()
o.backward()
```
The error is raised, because the non-differentiable path is included in dependency calculation (we only check `requires_grad` locally at the moment). It doesn't affect correctness and should never cause problems in most use cases. I'm putting this on hold until #662 is merged.

**Comment by danielamassiceti** [comment]
I ran into this issue as well. A temporary work around could be to re-package the tensor of argmax b into a new Variable.

Something like:
```
b = torch.max(a, 1)[1].repeat(1,5).float().data
o = (Variable(b) + a).sum()
```

This does not raise any error for me!

**Comment by apaszke** [comment]
Yeah, that's what I proposed on the forums. The issue was reported originally [in this thread](https://discuss.pytorch.org/t/runtime-error-caused-by-dependency-engine/465). Thanks for help though!

**Comment by apaszke** [comment]
Fixed in #752.
